library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
+onOneArgGen applyOperation typ arguments = do
+  -- length of arguments should be 1
+  -- so a reverse to mimic two should not do anything!
+  v <- traverse (protect . (inst >=> promoteTopStack)) arguments
+  -- see comments onTwoArgsGen
   case v of
     instr1 : _ -> do
-      res <-
-        if  | allConstants [val instr1] ->
-              let Env.Constant i1 = val instr1
-               in pure (f i1)
-            | otherwise -> do
-              addExpanded instr1
-              -- copyAndDrop 1
-              addInstr op
-              pure Env.Nop
-      -- remove when we normalize
+      res <- applyOperation instr1
       modify @"stack" (VStack.drop 1)
       consVal res typ
       pure res
     _ -> throw @"compilationError" Types.NotEnoughArguments
 
--- todo remove repeat pattern
--- Cons val here?
-pushConstant :: Env.Reduction m => Types.Type -> [Types.RawTerm] -> m Env.Expanded
-pushConstant typ instrs = do
-  v <- traverse (inst >=> promoteTopStack) instrs
-  res <- case v of
-    Env.Constant _ : _ ->
-      pure Env.Nop
-    instr1 : _ ->
-      pure instr1
-    _ -> throw @"compilationError" Types.NotEnoughArguments
-  modify @"stack" (VStack.drop 1)
-  consVal res typ
-  pure res
+------------------------------------------------------------
+-- Functions to call from appliers/reducers above
+------------------------------------------------------------
+
+onTwoArgsNoConst ::
+  Env.Reduction m => Instr.ExpandedOp -> Types.Type -> [Types.RawTerm] -> m Env.Expanded
+onTwoArgsNoConst op =
+  onTwoArgsGen (noTwoConstantCase op)
+
+onOneArgsNoConst ::
+  Env.Reduction m => Instr.ExpandedOp -> Types.Type -> [Types.RawTerm] -> m Env.Expanded
+onOneArgsNoConst op =
+  onOneArgGen (noOneConstantCase op)
+
+onTwoArgs :: OnTerm2 m (V.Value' Types.Op) Env.Expanded
+onTwoArgs op f =
+  onTwoArgsGen
+    ( \instr2 instr1 -> do
+        -- May be the wrong order?
+        let instrs = [instr2, instr1]
+        -- check if they are all constants
+        -- apply function if so
+        if  | allConstants (val <$> instrs) ->
+              let Env.Constant i1 = val instr1
+                  Env.Constant i2 = val instr2
+               in pure (f i1 i2)
+            | otherwise -> noTwoConstantCase op instr2 instr1
+    )
+
+onOneArgs :: OnTerm1 m (V.Value' Types.Op) Env.Expanded
+onOneArgs op f =
+  onOneArgGen
+    ( \instr1 -> do
+        -- check if they are all constants
+        -- apply function if so
+        if  | allConstants [val instr1] ->
+              let Env.Constant i1 = val instr1
+               in pure (f i1)
+            | otherwise -> noOneConstantCase op instr1
+    )
+
+------------------------------------------------------------
+-- Helper predicate/abstractions for the on functions
+------------------------------------------------------------
+
+noTwoConstantCase ::
+  Env.Ops m => Instr.ExpandedOp -> Protect -> Protect -> m Env.Expanded
+noTwoConstantCase op instr2 instr1 = do
+  -- May be the wrong order?
```

Comment:
this will result in the stack `res1 : res2 : ...` after execution presumably, is that what we want?

Reply:
Yes, that is what we want, at least that is what we've had in the past.